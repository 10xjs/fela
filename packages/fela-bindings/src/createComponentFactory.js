/* @flow */
import { combineRules } from 'fela'
import reduce from 'lodash/reduce'

import hoistStatics from './hoistStatics'
import extractPassThroughProps from './extractPassThroughProps'
import extractUsedProps from './extractUsedProps'
import resolvePassThrough from './resolvePassThrough'
import resolveUsedProps from './resolveUsedProps'

export default function createComponentFactory(
  createElement: Function,
  withTheme: Function,
  ProgressiveStyle: Function,
  contextTypes?: Object,
  withProxy: boolean = false,
  alwaysPassThroughProps: Array<string> = []
): Function {
  return function createComponent(
    rule: Function,
    type: any = 'div',
    passThroughProps: Array<string> | Function = []
  ): Function {
    const displayName = rule.name ? rule.name : 'FelaComponent'

    const FelaComponent = (
      {
        children,
        _felaTheme,
        _felaRule,
        extend,
        innerRef,
        id,
        style,
        as,
        className,
        passThrough = [],
        ...otherProps
      },
      { renderer }
    ) => {
      if (!renderer) {
        throw new Error(
          "createComponent() can't render styles without the renderer in the context. Missing react-fela's <Provider /> at the app root?"
        )
      }

      const usedProps = withProxy ? extractUsedProps(rule, _felaTheme) : []

      const rules = [rule]
      if (_felaRule) {
        rules.push(_felaRule)
      }
      if (extend) {
        typeof extend === 'function'
          ? rules.push(extend)
          : rules.push(() => extend)
      }
      const combinedRule = combineRules(...rules)

      // improve developer experience with monolithic renderer
      if (process.env.NODE_ENV !== 'production' && renderer.prettySelectors) {
        const componentName =
          typeof type === 'string' ? type : type.displayName || type.name || ''

        combinedRule.selectorPrefix = `${displayName}_${componentName}__`
      }
      // compose passThrough props from arrays or functions
      const resolvedPassThrough = [
        ...alwaysPassThroughProps,
        ...resolvePassThrough(passThroughProps, otherProps),
        ...resolvePassThrough(passThrough, otherProps),
        ...(withProxy ? resolveUsedProps(usedProps, otherProps) : []),
      ]

      const ruleProps = {
        ...otherProps,
        theme: _felaTheme,
        as,
        id,
      }

      // if the component renders into another Fela component
      // we pass down the combinedRule as well as both
      if (type._isFelaComponent) {
        return createElement(
          type,
          {
            _felaRule: combinedRule,
            passThrough: resolvedPassThrough,
            innerRef,
            style,
            className,
            as,
            id,
            ...otherProps,
          },
          children
        )
      }

      const componentProps = extractPassThroughProps(
        resolvedPassThrough,
        otherProps
      )

      if (id) {
        componentProps.id = id
      }

      if (innerRef) {
        componentProps.ref = innerRef
      }

      const previousCache = { ...renderer.cache }

      // fela-native support
      if (renderer.isNativeRenderer) {
        const felaStyle = renderer.renderRule(combinedRule, ruleProps)
        componentProps.style = style ? [style, felaStyle] : felaStyle
      } else {
        if (style) {
          componentProps.style = style
        }
        const cls = className ? `${className} ` : ''
        componentProps.className =
          cls + renderer.renderRule(combinedRule, ruleProps)
      }

      const customType = as || type
      const element = createElement(customType, componentProps, children)

      if (renderer.isProgressiveRenderer) {
        const cacheEntries = reduce(
          renderer.cache,
          (entries, value, key) => {
            if (!previousCache.hasOwnProperty(key)) {
              entries.push(value)
            }

            return entries
          },
          []
        )

        return [
          createElement(ProgressiveStyle, {
            cacheEntries,
            renderer
          }),
          element
        ]
      }

      return element
    }

    if (contextTypes) {
      FelaComponent.contextTypes = contextTypes
    }

    // use the rule name as display name to better debug with react inspector
    FelaComponent.displayName = displayName
    FelaComponent._isFelaComponent = true

    const themedComponent = withTheme(FelaComponent, '_felaTheme')
    return hoistStatics(themedComponent, type)
  }
}
