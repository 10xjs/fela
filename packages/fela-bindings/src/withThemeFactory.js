/* @flow */
import hoistStatics from './hoistStatics'

export default function withThemeFactory(
  createElement: Function,
  FelaTheme: Function
): Function {
  return function withTheme(
    component: Object,
    propName?: string = 'theme'
  ): Object {
    const WithTheme = props =>
      createElement(FelaTheme, undefined, theme =>
        createElement(component, {
          ...props,
          [propName]: theme,
        })
      )

    return hoistStatics(WithTheme, component)
  }
}
