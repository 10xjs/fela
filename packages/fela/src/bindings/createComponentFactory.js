/* @flow */
import {
  extractPassThroughProps,
  extractUsedProps,
  resolvePassThrough,
  resolveUsedProps
} from 'fela-utils'
import combineRules from '../combineRules'

export default function createComponentFactory(
  createElement: Function,
  contextTypes?: Object,
  withProxy: boolean = false
): Function {
  return function createComponent(
    rule: Function,
    type: any = 'div',
    passThroughProps: Array<string> | Function = []
  ): Function {
    const displayName = rule.name ? rule.name : 'FelaComponent'
    const usedProps = withProxy ? extractUsedProps(rule) : {}
    const defaultProps = type.defaultProps || {}

    const FelaComponent = (
      { children, _felaRule, passThrough = [], ...otherProps },
      { renderer, theme }
    ) => {
      if (!renderer) {
        throw new Error(
          "createComponent() can't render styles without the renderer in the context. Missing react-fela's <Provider /> at the app root?"
        )
      }

      const combinedRule = _felaRule ? combineRules(rule, _felaRule) : rule
      const ruleProps = {
        ...defaultProps,
        ...otherProps
      }

      // improve developer experience with monolithic renderer
      if (renderer.prettySelectors) {
        const componentName =
          typeof type === 'string' ? type : type.displayName || type.name || ''

        combinedRule.selectorPrefix = `${displayName}_${componentName}__`
      }
      // compose passThrough props from arrays or functions
      const resolvedPassThrough = [
        ...resolvePassThrough(passThroughProps, ruleProps),
        ...resolvePassThrough(passThrough, ruleProps),
        ...(withProxy ? resolveUsedProps(usedProps, ruleProps) : [])
      ]

      // if the component renders into another Fela component
      // we pass down the combinedRule as well as both
      if (type._isFelaComponent) {
        return createElement(
          type,
          {
            _felaRule: combinedRule,
            passThrough: resolvedPassThrough,
            ...ruleProps
          },
          children
        )
      }

      const componentProps = extractPassThroughProps(
        resolvedPassThrough,
        ruleProps
      )

      ruleProps.theme = theme || {}

      // fela-native support
      if (renderer.isNativeRenderer) {
        const felaStyle = renderer.renderRule(combinedRule, ruleProps)
        componentProps.style = ruleProps.style
          ? [ruleProps.style, felaStyle]
          : felaStyle
      } else {
        if (ruleProps.style) {
          componentProps.style = ruleProps.style
        }

        const cls = ruleProps.className ? `${ruleProps.className} ` : ''
        componentProps.className =
          cls + renderer.renderRule(combinedRule, ruleProps)
      }

      if (ruleProps.id) {
        componentProps.id = ruleProps.id
      }

      if (ruleProps.innerRef) {
        componentProps.ref = ruleProps.innerRef
      }

      const customType = ruleProps.is || type
      return createElement(customType, componentProps, children)
    }

    if (contextTypes) {
      FelaComponent.contextTypes = contextTypes
    }

    // use the rule name as display name to better debug with react inspector
    FelaComponent.displayName = displayName
    FelaComponent._isFelaComponent = true

    return FelaComponent
  }
}
