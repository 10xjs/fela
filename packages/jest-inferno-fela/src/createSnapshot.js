import { render } from 'inferno'
import { createElement } from 'inferno-create-element'
import { createRenderer } from 'fela'
import { Provider, ThemeProvider } from 'inferno-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  Provider,
  ThemeProvider
)
