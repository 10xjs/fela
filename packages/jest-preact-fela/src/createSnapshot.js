import { createElement, render } from 'preact'
import { createRenderer } from 'fela'
import { RendererProvider, ThemeProvider } from 'preact-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  RendererProvider,
  ThemeProvider
)
