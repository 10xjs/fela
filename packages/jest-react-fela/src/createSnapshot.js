import { createElement } from 'react'
import { render } from 'react-dom'
import { createRenderer } from 'fela'
import { Provider, ThemeProvider } from 'react-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  Provider,
  ThemeProvider
)
