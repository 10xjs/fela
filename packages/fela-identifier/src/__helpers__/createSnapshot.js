import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { render } from 'react-dom'
import { css } from 'js-beautify'

import RendererProviderFactory from '../../../fela-bindings/src/RendererProviderFactory'
import ThemeProviderFactory from '../../../fela-bindings/src/ThemeProviderFactory'
import { THEME_CHANNEL } from '../../../fela-bindings/src/themeChannel'

import renderToString from '../../../fela-tools/src/renderToString'
import createRenderer from '../../../fela/src/createRenderer'

const RendererProvider = RendererProviderFactory(
  Component,
  children => children,
  {
    childContextTypes: { renderer: PropTypes.object },
  }
)

const BaseThemeProvider = ThemeProviderFactory(
  Component,
  children => children,
  {
    propTypes: {
      theme: PropTypes.object.isRequired,
      overwrite: PropTypes.bool,
    },
    childContextTypes: {
      [THEME_CHANNEL]: PropTypes.object,
    },
    contextTypes: {
      [THEME_CHANNEL]: PropTypes.object,
    },
    defaultProps: {
      overwrite: false,
    },
  }
)

export default function createSnapshot(
  component,
  theme = {},
  renderer = createRenderer(),
  Provider = RendererProvider,
  ThemeProvider = BaseThemeProvider
) {
  const div = document.createElement('div')

  render(
    <Provider renderer={renderer}>
      <ThemeProvider theme={theme}>{component}</ThemeProvider>
    </Provider>,
    div
  )

  return {
    html: div.innerHTML,
    css: css(renderToString(renderer)),
  }
}
