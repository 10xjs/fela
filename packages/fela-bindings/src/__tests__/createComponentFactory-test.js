import 'raf/polyfill'
import React, { createElement, Component as BaseComponent } from 'react'
import PropTypes from 'prop-types'

import createComponentFactory from '../createComponentFactory'
import withThemeFactory from '../withThemeFactory'
import { THEME_CHANNEL } from '../themeChannel'

import createRenderer from '../../../fela/src/createRenderer'
import monolithic from '../../../fela-monolithic/src'

import createSnapshot from '../__helpers__/createSnapshot'

const withTheme = withThemeFactory(BaseComponent, createElement, {
  [THEME_CHANNEL]: PropTypes.object,
})

const createComponent = createComponentFactory(createElement, withTheme, {
  renderer: PropTypes.object,
})

const createComponentWithExtraPassThrough = createComponentFactory(
  createElement,
  withTheme,
  { renderer: PropTypes.object },
  false,
  ['testProp']
)

const createComponentWithProxy = createComponentFactory(
  createElement,
  withTheme,
  {
    renderer: PropTypes.object,
  },
  true
)

describe('Creating Components from Fela rules', () => {
  it('should return a Component', () => {
    const rule = props => ({
      color: props.color,
      fontSize: '16px',
    })

    const component = createComponent(rule)

    expect(component).toBeInstanceOf(Function)
  })

  it('should render fela rules depending on the passed props', () => {
    const rule = props => ({
      color: props.color,
      fontSize: '16px',
    })

    const Component = createComponent(rule)

    expect(createSnapshot(<Component color="black" />)).toMatchSnapshot()
  })

  it('should include defaultProps if provided', () => {
    const rule = props => ({
      color: props.color,
      fontSize: '16px',
    })

    const Comp = ({ color, className }) => (
      <div className={className}>{color}</div>
    )

    Comp.defaultProps = {
      color: 'red',
    }

    const Component = createComponent(rule, Comp)

    expect(createSnapshot(<Component />)).toMatchSnapshot()
  })

  it('should use the theme for static rendering by default', () => {
    const rule = props => ({
      color: props.theme.color,
      fontSize: '16px',
    })

    const Component = createComponent(rule)

    expect(createSnapshot(<Component />, { color: 'red' })).toMatchSnapshot()
  })

  it('should allow nested theme objects', () => {
    const rule = props => ({
      color: props.theme.header.color,
      fontSize: '16px',
    })

    const Component = createComponent(rule)

    expect(
      createSnapshot(<Component />, { header: { color: 'black' } })
    ).toMatchSnapshot()
  })

  it('should not pass props to the element', () => {
    const rule = props => ({
      color: props.color,
      fontSize: '16px',
    })

    const Component = createComponent(rule, 'div', ['onClick'])

    expect(
      createSnapshot(<Component data-foo="bar" data-bar="baz" />)
    ).toMatchSnapshot()
  })

  it('should pass all props to the element', () => {
    const rule = props => ({
      color: props.color,
      fontSize: '16px',
    })

    const Component = createComponent(rule, 'div', Object.keys)

    expect(
      createSnapshot(<Component data-foo="bar" data-bar="baz" />)
    ).toMatchSnapshot()
  })

  it('should pass special props to the component', () => {
    const rule = props => ({
      color: props.as === 'i' ? props.color : 'red',
      fontSize: '16px',
    })

    const Component = createComponent(rule)

    expect(createSnapshot(<Component />)).toMatchSnapshot()
    expect(createSnapshot(<Component as="i" />)).toMatchSnapshot()
  })

  it('should pass extended special props to the component', () => {
    const rule = props => ({
      color: props.as === 'i' ? props.color : 'red',
      fontSize: 16,
    })

    const UnderlyingComp = ({ testProp }) => (
      <div>{testProp ? 'testProp exists' : 'testProp does not exist'}</div>
    )
    const Component = createComponentWithExtraPassThrough(rule, UnderlyingComp)

    expect(
      createSnapshot(<Component color="blue" testProp={{ tabIndex: -1 }} />)
    ).toMatchSnapshot()
    expect(createSnapshot(<Component color="blue" />)).toMatchSnapshot()
  })

  it('should compose styles', () => {
    const rule = () => ({
      color: 'blue',
      fontSize: '16px',
    })

    const anotherRule = () => ({
      color: 'red',
      lineHeight: 1.2,
    })

    const Comp = createComponent(rule)
    const ComposedComp = createComponent(anotherRule, Comp)

    expect(createSnapshot(<ComposedComp />)).toMatchSnapshot()
  })

  it('should pass style, as, id, className and innerRef to composed components', () => {
    const rule = () => ({
      color: 'blue',
      fontSize: '16px',
    })

    const anotherRule = props => ({
      color: props.color,
      lineHeight: 1.2,
    })

    const Comp = createComponent(rule)
    const ComposedComp = createComponent(anotherRule, Comp)

    ComposedComp.defaultProps = {
      color: 'green',
    }

    expect(createSnapshot(<ComposedComp as="i" />)).toMatchSnapshot()
  })

  it('should extend the rule properties with an object', () => {
    const rule = () => ({
      color: 'blue',
      fontSize: '16px',
    })

    const Comp = createComponent(rule)

    expect(
      createSnapshot(
        <Comp extend={{ fontSize: '14px', backgroundColor: 'red' }} />
      )
    ).toMatchSnapshot()
  })

  it('should extend the rule properties with a function', () => {
    const rule = () => ({
      color: 'blue',
      fontSize: '16px',
    })

    const Comp = createComponent(rule)

    const extendRule = props => ({
      fontSize: '14px',
      backgroundColor: props.bgColor,
    })

    expect(
      createSnapshot(<Comp extend={extendRule} bgColor="red" />)
    ).toMatchSnapshot()
  })

  it('should compose passThrough props', () => {
    const Component = createComponent(() => ({}), 'div', Object.keys)
    const ComposedComponent = createComponent(() => ({}), Component, [
      'data-foo',
    ])

    expect(
      createSnapshot(<ComposedComponent data-foo="bar" color="red" />)
    ).toMatchSnapshot()
  })

  it('should only use the rule name as displayName', () => {
    const Button = () => ({
      color: 'red',
      fontSize: '16px',
    })

    const component = createComponent(Button)

    expect(component.displayName).toEqual('Button')
  })

  it('should use a dev-friendly className with monolithic renderer', () => {
    const Button = () => ({
      fontSize: '16px',
    })

    const Component = createComponent(Button)

    expect(
      createSnapshot(
        <Component />,
        {},
        createRenderer({
          enhancers: [
            monolithic({
              prettySelectors: true,
            }),
          ],
        })
      )
    ).toMatchSnapshot()
  })

  it('should use a dev-friendly className and the selectorPrefix', () => {
    const Button = () => ({
      fontSize: '16px',
    })

    const Component = createComponent(Button)

    expect(
      createSnapshot(
        <Component />,
        {},
        createRenderer({
          selectorPrefix: 'fela-',
          enhancers: [
            monolithic({
              prettySelectors: true,
            }),
          ],
        })
      )
    ).toMatchSnapshot()
  })

  it('should only use the rule name as displayName', () => {
    const Button = () => ({
      color: 'red',
      fontSize: '16px',
    })
    const Component = createComponent(Button)

    expect(createSnapshot(<Component />)).toMatchSnapshot()
  })

  it('should replace unallowed symbols in className with underscore', () => {
    const rule = () => ({
      fontSize: '16px',
    })

    const Parent = () => <div>Hello World</div>
    Parent.displayName = 'connect(Component)'
    const Component = createComponent(rule, Parent)

    expect(
      createSnapshot(
        <Component />,
        {},
        createRenderer({
          enhancers: [
            monolithic({
              prettySelectors: true,
            }),
          ],
        })
      )
    ).toMatchSnapshot()
  })

  it('should replace rare unallowed symbols in className with underscore', () => {
    const rule = () => ({
      fontSize: '16px',
    })

    const Parent = () => <div>Hello World</div>
    Parent.displayName = '1!@#$%^&*{}/=\\'
    const Component = createComponent(rule, Parent)

    expect(
      createSnapshot(
        <Component />,
        {},
        createRenderer({
          enhancers: [
            monolithic({
              prettySelectors: true,
            }),
          ],
        })
      )
    ).toMatchSnapshot()
  })
})

describe('Creating Components with a Proxy for props from Fela rules', () => {
  it('should not pass props used in rules to the element', () => {
    const rule = props => ({
      color: props.color,
      fontSize: '16px',
    })
    const Component = createComponentWithProxy(rule, 'div')

    expect(createSnapshot(<Component data-foo="bar" color />)).toMatchSnapshot()
  })

  it('should pass props used in rules specified in passThroughProps to the element', () => {
    const rule = props => ({
      color: props['data-color'],
      fontSize: '16px',
    })
    const Component = createComponentWithProxy(rule, 'div', ['data-color'])

    expect(
      createSnapshot(<Component data-foo="bar" data-color="blue" />)
    ).toMatchSnapshot()
  })

  it('should pass props except innerRef', () => {
    const rule = props => ({
      color: props.color,
      fontSize: '16px',
    })
    const Component = createComponentWithProxy(rule, 'div')

    expect(
      createSnapshot(<Component color="black" innerRef={() => 'test'} />)
    ).toMatchSnapshot()
  })
})
