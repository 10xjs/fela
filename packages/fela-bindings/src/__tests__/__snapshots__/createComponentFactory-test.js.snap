// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating Components from Fela rules should allow nested theme objects 1`] = `
".a {
  color: black;
}
.b {
  font-size: 16px;
}


<div className=a b />;
"
`;

exports[`Creating Components from Fela rules should compose passThrough props 1`] = `
"

<div data-foo=bar color=red className />;
"
`;

exports[`Creating Components from Fela rules should compose styles 1`] = `
".a {
  color: red;
}
.b {
  font-size: 16px;
}
.c {
  line-height: 1.2;
}


<div className=a b c />;
"
`;

exports[`Creating Components from Fela rules should extend the rule properties with a function 1`] = `
".a {
  color: blue;
}
.b {
  font-size: 14px;
}
.c {
  background-color: red;
}


<div className=a b c />;
"
`;

exports[`Creating Components from Fela rules should extend the rule properties with an object 1`] = `
".a {
  color: blue;
}
.b {
  font-size: 14px;
}
.c {
  background-color: red;
}


<div className=a b c />;
"
`;

exports[`Creating Components from Fela rules should include defaultProps if provided 1`] = `
".a {
  color: red;
}
.b {
  font-size: 16px;
}


<div className=a b>red</div>;
"
`;

exports[`Creating Components from Fela rules should not pass props to the element 1`] = `
".a {
  font-size: 16px;
}


<div className=a />;
"
`;

exports[`Creating Components from Fela rules should only use the rule name as displayName 1`] = `
".a {
  color: red;
}
.b {
  font-size: 16px;
}


<div className=a b />;
"
`;

exports[`Creating Components from Fela rules should pass all props to the element 1`] = `
".a {
  font-size: 16px;
}


<div data-foo=bar data-bar=baz className=a />;
"
`;

exports[`Creating Components from Fela rules should pass extended special props to the component 1`] = `
".a {
  color: red;
}
.b {
  font-size: 16;
}


<div>testProp exists</div>;
"
`;

exports[`Creating Components from Fela rules should pass extended special props to the component 2`] = `
".a {
  color: red;
}
.b {
  font-size: 16;
}


<div>testProp does not exist</div>;
"
`;

exports[`Creating Components from Fela rules should pass special props to the component 1`] = `
".a {
  color: red;
}
.b {
  font-size: 16px;
}


<div className=a b />;
"
`;

exports[`Creating Components from Fela rules should pass special props to the component 2`] = `
".a {
  font-size: 16px;
}


<i className=a />;
"
`;

exports[`Creating Components from Fela rules should pass style, as, id, className and innerRef to composed components 1`] = `
".a {
  color: green;
}
.b {
  font-size: 16px;
}
.c {
  line-height: 1.2;
}


<i className=a b c />;
"
`;

exports[`Creating Components from Fela rules should render fela rules depending on the passed props 1`] = `
".a {
  color: black;
}
.b {
  font-size: 16px;
}


<div className=a b />;
"
`;

exports[`Creating Components from Fela rules should replace rare unallowed symbols in className with underscore 1`] = `
".rule_1_______________3vb6hs {
  font-size: 16px;
}


<div>Hello World</div>;
"
`;

exports[`Creating Components from Fela rules should replace unallowed symbols in className with underscore 1`] = `
".rule_connect_Component___3vb6hs {
  font-size: 16px;
}


<div>Hello World</div>;
"
`;

exports[`Creating Components from Fela rules should use a dev-friendly className and the selectorPrefix 1`] = `
".fela-Button_div__3vb6hs {
  font-size: 16px;
}


<div className=fela-Button_div__3vb6hs />;
"
`;

exports[`Creating Components from Fela rules should use a dev-friendly className with monolithic renderer 1`] = `
".Button_div__3vb6hs {
  font-size: 16px;
}


<div className=Button_div__3vb6hs />;
"
`;

exports[`Creating Components from Fela rules should use the theme for static rendering by default 1`] = `
".a {
  color: red;
}
.b {
  font-size: 16px;
}


<div className=a b />;
"
`;

exports[`Creating Components with a Proxy for props from Fela rules should not pass props used in rules to the element 1`] = `
".a {
  color: true;
}
.b {
  font-size: 16px;
}


<div data-foo=bar className=a b />;
"
`;

exports[`Creating Components with a Proxy for props from Fela rules should pass props except innerRef 1`] = `
".a {
  color: black;
}
.b {
  font-size: 16px;
}


<div className=a b />;
"
`;

exports[`Creating Components with a Proxy for props from Fela rules should pass props used in rules specified in passThroughProps to the element 1`] = `
".a {
  color: blue;
}
.b {
  font-size: 16px;
}


<div data-color=blue data-foo=bar className=a b />;
"
`;
